docs/csharp/proto2.md

```diff
@@ -0,0 +1,184 @@
+
+var foo = new Foo();
+foo.SetExtension(FooExt, new Baz());
+foo.GetOrInitializeExtension(RepeatedFooExt).Add(new Baz());
+```
+
+# APIs
+
+### Message initialization
+
+Checking message initialization is not handled automatically by the library, but can be done manually via the
+`IsInitialized` extension method in `MessageExtensions`. Please note, parsers and input streams don't check messages 
+for initialization on their own and throw errors. Instead it's up to you to handle messages with missing required fields
+in whatever way you see fit.
+
+### Extension registries
+
+Just like in Java, extension registries can be constructed to parse extensions when reading new messages
+from input streams. The API is fairly similar to the Java API with some added bonuses with C# syntax sugars.
+
+```proto
+message Baz {
+    extend Foo {
+        optional Baz foo_ext = 124;
+    }
+}
+```
+```cs
+var registry = new ExtensionRegistry() 
+{ 
+    Baz.Extensions.FooExt
+};
+var foo = Foo.Factory.WithExtensionRegistry(registry).ParseFrom(input);
+Assert.True(foo.HasExtension(Bas.Extensions.FooExt));
+var fooNoRegistry = Foo.Factory.ParseFrom(input);
+Assert.False(foo.HasExtension(Bas.Extensions.FooExt));
+```
+
+### Custom options
+
+The original `CustomOptions` APIs are now deprecated. Using the new generated extension identifiers, 
+you can access extensions safely through the GetOption APIs. Note that cloneable values such as 
+repeated fields and messages will be deep cloned.
+
+Example based on custom options usage example [here](https://github.com/protocolbuffers/protobuf/issues/5007#issuecomment-411604515).
+```cs
+foreach (var service in input.Services)
+{
+    Console.WriteLine($"  {service.Name}");
+    foreach (var method in service.Methods)
+    {
+        var rule = method.GetOption(AnnotationsExtensions.Http);
+        if (rule != null)
+        {
+            Console.WriteLine($"    {method.Name}: {rule}");
+        }
+        else
+        {
+            Console.WriteLine($"    {method.Name}: no HTTP binding");
+        }
+    }
+}
+```
+
+### Reflection
+
+Reflection APIs have been included to access the new portions of the library.
+
+ * FieldDescriptor.Extension
+    * Gets the extension identifier behind an extension field, allowing it to be added to an ExtensionRegistry
+ * FieldDescriptor.IsExtension
+    * Returns whether a field is an extension of another type.
+ * FieldDescriptor.ExtendeeType
+    * Returns the extended type of an extension field
+ * IFieldAccessor.HasValue
+    * Returns whether a field's value is set. For proto3 fields, throws an InvalidOperationException.
+ * FileDescriptor.Syntax
+    * Gets the syntax of a file
+ * FileDescriptor.Extensions
+    * An immutable list of extensions defined in the file
+ * MessageDescriptor.Extensions
+    * An immutable list of extensions defined in the message
+
+```cs
+var extensions = Baz.Descriptor.Extensions.GetExtensionsInDeclarationOrder(Foo.Descriptor);
```

Comment:
idea: in the future, if this is a pattern that's used a lot, there could be an API (extension method) that allows you to create extension registry populated with all extensions from given descriptor?

Reply:
It seems easy enough to do manually via `AddRange` and `UnorderedExtensions`

```cs
var registry = new ExtensionRegistry();
registry.AddRange(Baz.Descriptor.Extensions.UnorderedExtensions);
```